include/range/v3/utility/semiregular_box.hpp

```diff
@@ -292,6 +292,17 @@ namespace ranges
         meta::if_c<IsConst || std::is_empty<T>::value, T, reference_wrapper<T>>,
         reference_wrapper<
             meta::if_c<IsConst, semiregular_box<T> const, semiregular_box<T>>>>;
+
+    namespace detail {
+        // For those range adapters that store a function by way of
+        // semiregular_box_ref_or_val_t if that alias is actually a T (that is,
+        // it owns the callable), then the range can be a safe_range since the
+        // adapter will not depend on the outer range surviving.
+        // This is a necessary, but not sufficient, condition
+        template <typename T>
+        RANGES_INLINE_VAR constexpr bool is_safe_fun =
+            defer::same_as<T, semiregular_box_ref_or_val_t<T>>;
+    }
```

Comment:
I would prefer not to consider these safe ranges at all. I want the freedom to move the functions into the view objects when compiler optimizers get good enough to eliminate the abstraction cost.

Reply:
I rolled all of these back - no more `is_safe_fun` or any conditions that depend on them.

However, I kept three views as conditionally safe (er, I guess borrowed now) implicitly based  n this: `keys`, `views`, and `zip`. On the basis that these are commonly used views, their function objects are empty so I know range-v3 stores them in the iterators, and it's useful to have more ranges be safe in this regard.

Reply:
Thanks, Barry.